src/main/java/net/openhft/chronicle/core/io/VanillaReferenceCounted.java

```diff
@@ -124,20 +123,21 @@ public void callOnRelease() throws ClosedIllegalStateException {
 
     @Override
     public void releaseLast(ReferenceOwner id) throws IllegalStateException {
-        for (; ; ) {
-            int v = value;
-            if (v <= 0) {
-                if (Jvm.supportThread())
-                    break;
-                throw newReleasedClosedIllegalStateException();
-            }
-            if (v > 1) {
-                throw new IllegalStateException(type.getName() + " not the last released");
-            }
-            if (valueCompareAndSet(1, 0)) {
-                callOnRelease();
-                break;
+        Exception thrownException = null;
+        try {
+            release(id);
+        } catch (Exception e) {
+            thrownException = e;
+        }
+        if (value > 0) {
+            final IllegalStateException ise = new IllegalStateException(type.getName() + " not the last released");
+            if (thrownException != null) {
+                ise.addSuppressed(thrownException);
             }
+            throw ise;
+        }
+        if (thrownException != null) {
+            Jvm.rethrow(thrownException);
```

Comment:
As discussed this behaviour was changed to match that of `TracingReferenceCounted`, i.e. when the reference released is not the last, it will be released then an exception will be thrown.